internal/services/vehicle_event_listener.go

```diff
@@ -13,119 +13,140 @@ import (
 }
 
-type VehicleEventData struct {
-	VehicleID string `json:"vehicleId"`
-	Odometer  int    `json:"odometer"`
-	// More fields later
+// SignalListener consumes signals from Kafka and processes them
+type SignalListener struct {
+	log          zerolog.Logger
+	webhookCache *WebhookCache
 }
 
-type VehicleEventListener struct {
-	log zerolog.Logger
-}
-
-func NewVehicleEventListener(logger zerolog.Logger) *VehicleEventListener {
-	return &VehicleEventListener{
-		log: logger,
+// NewSignalListener initializes a listener with a logger and an in-memory cache of webhooks
+func NewSignalListener(logger zerolog.Logger, wc *WebhookCache) *SignalListener {
+	return &SignalListener{
+		log:          logger,
+		webhookCache: wc,
 	}
 }
 
-func (l *VehicleEventListener) ProcessVehicleEvents(messages <-chan *message.Message) {
+// ProcessSignals is the entry point for Watermill. It loops over incoming messages
+func (l *SignalListener) ProcessSignals(messages <-chan *message.Message) {
 	for msg := range messages {
 		if err := l.processMessage(msg); err != nil {
-			l.log.Err(err).Msg("error processing vehicle event message")
+			l.log.Err(err).Msg("error processing signal message")
 		}
 		msg.Ack()
 	}
 }
 
-func (l *VehicleEventListener) processMessage(msg *message.Message) error {
+// processMessage parses the Signal, looks up matching webhooks, and evaluates conditions
+func (l *SignalListener) processMessage(msg *message.Message) error {
 	l.log.Info().
 		Str("message_id", msg.UUID).
 		Msgf("Received payload: %s", string(msg.Payload))
 
-	var evt CloudEvent[VehicleEventData]
-	if err := json.Unmarshal(msg.Payload, &evt); err != nil {
-		return errors.Wrap(err, "failed to parse vehicle event JSON")
+	var signal Signal
+	if err := json.Unmarshal(msg.Payload, &signal); err != nil {
+		return errors.Wrap(err, "failed to parse vehicle signal JSON")
 	}
 
 	l.log.Info().
-		Str("cloud_event_id", evt.ID).
-		Str("vehicle_id", evt.Data.VehicleID).
-		Int("odometer", evt.Data.Odometer).
-		Msg("Parsed VehicleEventData")
-
-	shouldTrigger, err := l.evaluateCEL(evt.Data.Odometer)
-	if err != nil {
-		return errors.Wrap(err, "CEL evaluation failed")
+		Uint32("token_id", signal.TokenID).
+		Str("signal_name", signal.Name).
+		Float64("value_number", signal.ValueNumber).
+		Str("value_string", signal.ValueString).
+		Msg("Parsed Signal")
+
+	webhooks := l.webhookCache.GetWebhooks(signal.TokenID, signal.Name)
+	if len(webhooks) == 0 {
+		return nil
 	}
-	if shouldTrigger {
-		l.log.Info().Msg("CEL condition met (odometer > 100): Webhook should be fired")
-		// Simulate a webhook call to a local endpoint.
-		if err := l.sendWebhookNotification(evt); err != nil {
-			l.log.Err(err).Msg("failed to send webhook notification")
+
+	for _, wh := range webhooks {
+		shouldFire, err := l.evaluateCondition(wh.Condition, &signal)
+		if err != nil {
+			l.log.Error().Err(err).Msg("failed to evaluate CEL condition")
+			continue
+		}
+		if shouldFire {
```

Comment:
there should also be a cooldown, probably before evaluating the condition, where you check if this event has fired for this tokenId in the last X period (i think we had it configurable in the db?)